vllm/transformers_utils/detokenizer.py

```diff
@@ -0,0 +1,178 @@
+
+        Args:
+            seq: The sequence to decode.
+            prms: The sampling parameters used to generate the sequence.
+            logprobs: The logprobs to decode.
+            all_input_ids: The token IDs of the sequence, including the
+                token generated this iteration.
+            generated_token_id_and_text: The token ID and text of the token
+                generated this iteration. If provided, will be used to
+                avoid decoding the token again.
+        """
+        if not logprobs:
+            return
+        tokenizer = self.get_tokenizer_for_seq(seq)
+        previous_tokens = all_input_ids[:-1]
+        for token_id, sample_logprob in logprobs.items():
+            # If the token was generated this iteration, use the provided text.
+            if (generated_token_id_and_text is not None
+                    and token_id == generated_token_id_and_text[0]):
+                sample_logprob.decoded_token = generated_token_id_and_text[1]
+                continue
+
+            if (sample_logprob.decoded_token is None and token_id != -1):
+                all_input_ids_with_logprob = previous_tokens + [token_id]
+                (_, new_text, _, _) = detokenize_incrementally(
+                    tokenizer=tokenizer,
+                    all_input_ids=all_input_ids_with_logprob,
+                    prev_tokens=seq.tokens,
+                    prefix_offset=seq.prefix_offset,
+                    read_offset=seq.read_offset,
+                    skip_special_tokens=prms.skip_special_tokens,
+                    spaces_between_special_tokens=prms.
+                    spaces_between_special_tokens,
+                )
+                sample_logprob.decoded_token = new_text
+
+    def decode_prompt_logprobs(
+            self, seq_group: SequenceGroup,
+            prompt_logprobs: List[Dict[int,
+                                       Logprob]]) -> List[Dict[int, Logprob]]:
+        """Decodes the logprobs for the prompt of a sequence group.
+
+        Args:
+            seq_group: The sequence group to decode.
+            prompt_logprobs: The logprobs to decode.
+        
+        Returns:
+            The prompt logprobs with the decoded tokens.
+        """
+        prms = seq_group.sampling_params
+        # We can pick any sequence for the prompt.
+        seq = next(iter(seq_group.seqs_dict.values()))
+        # Only prompt, without the generated token.
+        all_token_ids = seq.get_token_ids()
+        prompt_token_ids = all_token_ids[:-1]
+        tokenizer = self.get_tokenizer_for_seq(seq)
+        prefix_offset = 0
+        read_offset = 0
+        next_iter_prefix_offset = 0
+        next_iter_read_offset = 0
+        next_iter_tokens = []
+        prev_tokens = None
+
+        for token_position, prompt_logprobs_for_token in enumerate(
+                prompt_logprobs):
+            if not prompt_logprobs_for_token:
```

Comment:
this is the case of empty dict? if so please just use `len`, took sometimes to figure this out

Reply:
it can be None!

Reply:
Ah in that case change type hint add `Optional`